File: 29. Divide Two Integers.cpp
class Solution {
public:
    int divide(int dividend, int divisor) {
        int sign = dividend < 0 ^ divisor < 0 ? -1 : 1;
        long long n = abs((long long) dividend), m = abs((long long) divisor);
        long long q = 0;

        for (long long t = 0, i = 31; i >= 0; i--) {
            if (t + (m << i) <= n) {
                t += m << i, q |= 1LL << i;
            }
        }

        if (sign < 0) q = -q;
        return q >= INT_MAX ? INT_MAX : q <= INT_MIN ? INT_MIN : q;
    }
};

File: java/29. Divide Two Integers.java
// O(logn)
class Solution {
    public int divide(int dividend, int divisor) {
        int sign = 1;
        if ((dividend < 0) ^ (divisor < 0)) {
            sign = -1;
        }
        long num = Math.abs((long) dividend);
        long den = Math.abs((long) divisor);
        int times = 1;
        long res = 0;
        while ((den << 1) < num) {
            times <<= 1;
            den <<= 1;
        }
        while (times > 0) {
            while (num >= den) {
                res += times;
                num -= den;
            }
            den >>>= 1;
            times >>>= 1;
        }
        res *= sign;
        return res > Integer.MAX_VALUE ? Integer.MAX_VALUE : (int)res;
    }
}
